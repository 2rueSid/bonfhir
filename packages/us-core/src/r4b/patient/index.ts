export * from "./birth-sex";
export * from "./gender-identity";
export * from "./patient";
export * from "./race";
export * from "./sex";
