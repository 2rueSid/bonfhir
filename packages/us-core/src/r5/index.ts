export * from "./direct-email";
export * from "./jurisdiction";
export * from "./patient";
export * from "./questionnaire-response";
