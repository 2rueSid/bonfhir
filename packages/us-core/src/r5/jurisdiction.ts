/**
 * https://www.hl7.org/fhir/us/core/StructureDefinition-us-core-jurisdiction.html
 */
export const US_CORE_JURISDICTION_URL =
  "http://hl7.org/fhir/us/core/StructureDefinition/us-core-jurisdiction";
