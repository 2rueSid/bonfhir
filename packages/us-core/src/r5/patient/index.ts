export * from "./birth-sex";
export * from "./ethnicity";
export * from "./gender-identity";
export * from "./patient";
export * from "./race";
export * from "./sex";
export * from "./tribal-affiliation";
