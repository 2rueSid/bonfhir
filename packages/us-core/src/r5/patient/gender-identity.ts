import { extension } from "@bonfhir/core/r5";

/**
 * https://hl7.org/fhir/us/core/StructureDefinition-us-core-genderIdentity.html
 */
export const US_CORE_GENDER_IDENTITY_URL =
  "http://hl7.org/fhir/us/core/StructureDefinition/us-core-genderIdentity";

/**
 * https://hl7.org/fhir/us/core/StructureDefinition-us-core-genderIdentity.html
 */
export function usCorePatientGenderIdentity() {
  return extension({
    kind: "valueCodeableConcept",
    url: US_CORE_GENDER_IDENTITY_URL,
  });
}
