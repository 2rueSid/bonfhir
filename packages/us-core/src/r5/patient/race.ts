/* eslint-disable @typescript-eslint/no-explicit-any */
import { Coding, HasExtension, SpecialExtension } from "@bonfhir/core/r5";

/**
 * https://hl7.org/fhir/us/core/StructureDefinition-us-core-race.html
 */
export const US_CORE_RACE_URL =
  "http://hl7.org/fhir/us/core/StructureDefinition/us-core-race";

/**
 * https://hl7.org/fhir/us/core/StructureDefinition-us-core-race.html
 */
export interface USCoreRace {
  /**
   * The 5 race category codes according to the [OMB Standards for Maintaining, Collecting, and Presenting Federal Data on Race and Ethnicity, Statistical Policy Directive No. 15, as revised, October 30, 1997](https://www.govinfo.gov/content/pkg/FR-1997-10-30/pdf/97-28653.pdf).
   */
  ombCategory?: Coding[] | undefined;

  /**
   * The 900+ CDC race codes that are grouped under one of the 5 OMB race category codes:.
   */
  detailed?: Coding[] | undefined;

  /**
   * Plain text representation of the race concept(s).
   */
  text: string;
}

/**
 * https://hl7.org/fhir/us/core/StructureDefinition-us-core-race.html
 */
export function usCoreRace(): USCoreRace | undefined {
  return {
    __isSpecialExtension: true,
    __get(target: HasExtension): USCoreRace | undefined {
      const extensionValue = (target.extension || []).find(
        (x) => x.url === US_CORE_RACE_URL,
      );

      if (!extensionValue) {
        return;
      }

      return {
        ombCategory:
          extensionValue.extension
            ?.filter((x) => x.url === "ombCategory" && x.valueCoding)
            ?.map((x) => x.valueCoding!) || [],
        detailed:
          extensionValue.extension
            ?.filter((x) => x.url === "detailed" && x.valueCoding)
            ?.map((x) => x.valueCoding!) || [],
        text:
          extensionValue.extension?.find((x) => x.url === "text")
            ?.valueString || "",
      };
    },
    __set(
      target: HasExtension,
      value: USCoreRace | undefined,
    ): USCoreRace | undefined {
      if (!value) {
        target.extension = target.extension?.filter(
          (x) => x.url !== US_CORE_RACE_URL,
        );

        if (target.extension?.length === 0) {
          target.extension = undefined;
        }
        return;
      }

      const newExtensionValue = {
        url: US_CORE_RACE_URL,
        extension: [
          ...(value.ombCategory || []).map((valueCoding) => ({
            url: "ombCategory",
            valueCoding,
          })),
          ...(value.detailed || []).map((valueCoding) => ({
            url: "detailed",
            valueCoding,
          })),
          {
            url: "text",
            valueString: value.text,
          },
        ],
      };

      target.extension = [
        ...(target.extension?.filter((x) => x.url !== US_CORE_RACE_URL) || []),
        newExtensionValue,
      ];

      return value;
    },
  } satisfies SpecialExtension as any;
}
