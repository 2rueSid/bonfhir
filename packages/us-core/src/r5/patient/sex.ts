import { extension } from "@bonfhir/core/r5";

/**
 * https://hl7.org/fhir/us/core/StructureDefinition-us-core-sex.html
 */
export const US_CORE_SEX_URL =
  "http://hl7.org/fhir/us/core/StructureDefinition/us-core-sex";

/**
 * https://vsac.nlm.nih.gov/valueset/2.16.840.1.113762.1.4.1240.3/expansion
 */
export type UsCoreSex =
  | "184115007"
  | "248152002"
  | "248153007"
  | "asked-declined";

/**
 * https://hl7.org/fhir/us/core/StructureDefinition-us-core-sex.html
 *
 * https://vsac.nlm.nih.gov/valueset/2.16.840.1.113762.1.4.1240.3/expansion
 */
export function usCorePatientSex(): UsCoreSex | undefined {
  return extension({
    kind: "valueCode",
    url: US_CORE_SEX_URL,
  }) as UsCoreSex | undefined;
}
