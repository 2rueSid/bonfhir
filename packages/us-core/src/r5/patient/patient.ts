import { usCorePatientBirthSex } from "./birth-sex";
import { usCorePatientGenderIdentity } from "./gender-identity";
import { usCoreRace } from "./race";
import { usCorePatientSex } from "./sex";

export function usCorePatient() {
  return {
    race: usCoreRace(),
    birthsex: usCorePatientBirthSex(),
    sex: usCorePatientSex(),
    genderIdentity: usCorePatientGenderIdentity(),
  };
}
