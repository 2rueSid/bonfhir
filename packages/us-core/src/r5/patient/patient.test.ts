import { codeableConcept, extendResource } from "@bonfhir/core/r5";
import { usCorePatient } from "./patient";

describe("patient", () => {
  const USCorePatient = extendResource("Patient", {
    ...usCorePatient(),
  });

  it("should work", () => {
    const patient = new USCorePatient();
    patient.race = {
      ombCategory: [
        {
          system: "urn:oid:2.16.840.1.113883.6.238",
          code: "2106-3",
          display: "White",
        },
        {
          system: "urn:oid:2.16.840.1.113883.6.238",
          code: "2028-9",
          display: "Asian",
        },
      ],
      detailed: [
        {
          system: "urn:oid:2.16.840.1.113883.6.238",
          code: "1586-7",
          display: "Shoshone",
        },
        {
          system: "urn:oid:2.16.840.1.113883.6.238",
          code: "2036-2",
          display: "Filipino",
        },
      ],
      text: "Mixed",
    };

    expect(patient.race).toMatchObject({
      ombCategory: [
        {
          system: "urn:oid:2.16.840.1.113883.6.238",
          code: "2106-3",
          display: "White",
        },
        {
          system: "urn:oid:2.16.840.1.113883.6.238",
          code: "2028-9",
          display: "Asian",
        },
      ],
      detailed: [
        {
          system: "urn:oid:2.16.840.1.113883.6.238",
          code: "1586-7",
          display: "Shoshone",
        },
        {
          system: "urn:oid:2.16.840.1.113883.6.238",
          code: "2036-2",
          display: "Filipino",
        },
      ],
      text: "Mixed",
    });

    patient.birthsex = "F";
    patient.sex = "184115007";

    patient.genderIdentity = codeableConcept({
      system: "http://terminology.hl7.org/CodeSystem/v3-NullFlavor",
      code: "UNK",
      display: "Unknown",
    });

    expect(patient.genderIdentity).toMatchObject(
      codeableConcept({
        system: "http://terminology.hl7.org/CodeSystem/v3-NullFlavor",
        code: "UNK",
        display: "Unknown",
      }),
    );

    expect(JSON.stringify(patient)).toMatchSnapshot();
  });
});
