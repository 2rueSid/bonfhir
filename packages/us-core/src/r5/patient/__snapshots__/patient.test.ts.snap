// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`patient should work 1`] = `"{"resourceType":"Patient","extension":[{"url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-race","extension":[{"url":"ombCategory","valueCoding":{"system":"urn:oid:2.16.840.1.113883.6.238","code":"2106-3","display":"White"}},{"url":"ombCategory","valueCoding":{"system":"urn:oid:2.16.840.1.113883.6.238","code":"2028-9","display":"Asian"}},{"url":"detailed","valueCoding":{"system":"urn:oid:2.16.840.1.113883.6.238","code":"1586-7","display":"Shoshone"}},{"url":"detailed","valueCoding":{"system":"urn:oid:2.16.840.1.113883.6.238","code":"2036-2","display":"Filipino"}},{"url":"text","valueString":"Mixed"}]},{"url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-birthsex","valueCode":"F"},{"url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-sex","valueCode":"184115007"},{"url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-genderIdentity","valueCodeableConcept":{"coding":[{"system":"http://terminology.hl7.org/CodeSystem/v3-NullFlavor","code":"UNK","display":"Unknown"}],"text":"Unknown"}}],"text":{"status":"generated","div":"<div xmlns=\\"http://www.w3.org/1999/xhtml\\"><ul></ul></div>"}}"`;
